src/red_black_tree.rs

```diff
@@ -735,6 +735,32 @@ impl<
             terminated: false,
         }
     }
+
+    pub fn filter<'a, P: for<'p> Fn(&'p K, &'p V) -> bool>(
+        &'a mut self,
+        predicate: P,
+    ) -> Vec<(K, V)> {
+        self.drain_filter(predicate).collect::<Vec<_>>()
+    }
+
+    pub fn drain_filter<'a, P: for<'p> Fn(&'p K, &'p V) -> bool>(
+        &'a mut self,
+        predicate: P,
+    ) -> RedBlackTreeDrainFilter<'a, K, V, P, MAX_SIZE> {
+        let node = self.root;
+        RedBlackTreeDrainFilter {
+            tree: self,
+            fwd_stack: vec![],
+            fwd_ptr: node,
+            fwd_node: None,
+            _rev_stack: vec![],
+            _rev_ptr: node,
+            rev_node: None,
+            terminated: false,
+            remove: vec![],
```

Comment:
An even smaller optimization to save heap would be the optional ability to pass in a mutable reference to `remove`